### ISSUE TITLE
Bug: TSTypeAnnotation erroneously contains its punctuator

### ISSUE THREAD

**Issue by zachbryant** [typescript-eslint/typescript-eslint#7727]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&showAST=es&fileType=.tsx&code=LAKALgngDgpgBAFWjATHAvHA3qOcBmA9oQPwBccARsQDYwCGAdgNygC%2BooMjArgLZwAymABOAS0YBzAKK8BOEHgBixDHADkSgPJb1AGnaggA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=true

### Repro Code

```typescript
type Type2 = {
  foo?: boolean;
}

enum StringEnum {
  Foo = 'FOO',
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {}
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expected the `loc` of interface keys to stop at the end of the type, before the semicolon. In my example, `loc.end.column` should be 15.

### Actual Result

`loc.end.column` is 16. Supposing `node` is the node whose getText is `foo?: boolean`, if you try to filter for the next punctuator with something like below, you'll end up with the punctuation from a completely different type or enum definition. Using this code, I get the comma from after `Foo` in the example enum.

```ts
const punctuator = sourceCode.getTokenAfter(node, {
    filter: n => (
        n.type === AST_TOKEN_TYPES.Punctuator &&
        new RegExp(`^[,;]$`).test(n.value)
      ),
    includeComments: false,
  })
```
